Proj1 citiroc checkbox

Joarfo opened the pull request:
Made changes according to comments on branch proj1_multiple_asic_cfg.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +568,50 @@ private void button_sendSC_Click(object sender, EventArgs e)
             {
+                if (sc_sendToNVM == 1)
+                {
+                    if (connectStatus == 1)
```
No need to make this check here; `sendSC()` -- if you bring back the code you commented out at the top of that method -- should make the check for you.

Joarfo replied:
This was added in response to a comment on the previous PR that suggested that the user should only be prompted for inputs when the board is connected. I.e the user should not go through the process of inputting a configuration number to then find out that the board was not connected. Should I remove it, but it somewhere else or leave it as is?

tstana replied:
Right, good point. You could just do:

```C#
if ((checkBox_sendToNVM.Checked == true) && (connectStatus != 1)
{
    button_sendSC.BackColor = Color.LightCoral;
    tmrButtonColor.Enabled = true;
    label_help.Text = "Please configure your connection.";
    return;
}
```

The timer is enabled so it can revert the color of the button and the `return` at the end is needed so you don't go through the rest of the method.

Joarfo replied:
Changed the connection check according to suggestion.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -667,8 +711,14 @@ private bool sendSC(int usbDevId, string strSC)
                 {
-                    protoCubes.SendCommand(ProtoCubesSerial.Command.SendCitirocConf,
-                        bytSC);
+                    if (sc_sendToNVM != 1)
```
You can use the `Checked` member of the new checkbox here.

tstana commented on CitirocUI/NVMConfigNumSelect.cs:
```diff
@@ -0,0 +1,59 @@
+    public class NVMVConfigNumSelect : Form
+    {
+
+        public static DialogResult InputBox(string title, string promptText, ref string value)
```
As outlined in `SlowControl.cs`, the `ref string value` could be made a `ref byte value`. I'm not sure if that can be done with a `DialogResult`, but please check.

Joarfo replied:
I will look into if this can be done, but I am also thinking that it might be better to change the name of the class entirely. In essence, what it does is it brings up a message box for typing in inputs; this type of user interface could be useful for other UI elements in the future and is not limited to only this situation. Changing its name could add this option for future functions. I was surprised when I didn't see this form interface in the options for the existing function (it probably exists, but I have not found it). What do you think?

tstana replied:
That could be a good idea, yes. However, if we are to make the check that the config ID is not larger than 255 here, it makes sense to have the dedicated form.

Joarfo replied:
Changed the name to "InputForm" to better reflect functionality and to add options for using the form for later. Was unable to properly fix value checks inside the form and had to move on to manage other works. The function has to return a string, not a byte, to properly interface with sendSC().

tstana replied:
OK. Will test and resolve once it's fixed.

Joarfo replied:
Fixed value and type check inside InputForm().

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +568,50 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                    {
+                        label_help.Text = "Please configure your connection.";
+                    }
+                }
```
Considering my comments above, here's how I would see the `if` statement above (disclaimer: it may not compile):

```C#
byte configNo;

if (checkBox_sendToNVM.Checked == true)
{
  if (NVMVConfigNumSelect.InputBox("Select Config Number", "Config Number:", ref configNo) == DialogResult.OK)
  {
    // `configNo` is assumed to be a `byte` here
    string n = Convert.ToString(configNo, 2);
    strSC += strRev(n);
  }
}
```

From here on out, `strSC` should be the right size for sending with whichever command within `sendSC()`.

tstana reviewed (changes requested):
@Joarfo, OK, we're getting there -- thanks for your changes. Please see my inline comments for some improvements that can be made. As usual, feel free to ask any questions.

Finally (sorry for not having replied to [your thread in the other MR](https://github.com/tstana/CitirocUI/pull/13#discussion_r776438022)), to send the `SelectNVMCitirocConf` command, I think the _Firmware_ tab would be the best place to add a textbox and a button for selecting the configuration. You could also envision only having a button and bringing up your NVM Conf Number Select form...

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +568,50 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                                else
+                                {
+                                    configNo = Convert.ToString(Convert.ToInt32(configNo, 10), 2);
+                                    configNo = configNo.PadLeft(8, '0');
```
The Padleft() is necessary since convert operation above for some reason removes leading zeroes. It may be that other convert commands could be used to get a proper 8-bit number with leading zeroes, but I have yet to find it.

Joarfo commented on CitirocUI/Citiroc.Designer.cs:
```diff
@@ -1006,19 +1007,6 @@ private void InitializeComponent()
-            this.label_weerocDotCom.TabIndex = 23;
-            this.label_weerocDotCom.Text = "www.weeroc.com";
-            this.label_weerocDotCom.TextAlign = System.Drawing.ContentAlignment.MiddleCenter;
-            // 
```
Readded this label

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +568,50 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                                    "Error",
+                                    MessageBoxButtons.OK,
+                                    MessageBoxIcon.Error);
+                                }
```
Was unable to find a solution to have this check inside the form without also breaking its functionality. Due to other deadlines and courses, finding this solution has been down prioritized and dropped for the time being.

tstana replied:
OK, I can make the change for you once I'll have access to the hardware.

Joarfo replied:
Ok I see now what you mean, I was thinking that ideally the form should remain open while the input was of incorrect size or type, and was unable to get that to work. I added the check to the form now.

tstana replied:
Thanks for adding the check in the form. Closing this thread, refer to the comments in the code of the form for hints on how to keep the user in the form for as long as they are typing a wrong value.

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -591,8 +635,8 @@ private void button_sendSC_Click(object sender, EventArgs e)
-            if (connectStatus != 1)
-                return false;
+            //if (connectStatus != 1)
+            //    return false;
```
Thanks for pointing this out, forgot to uncomment after running some tests. The check has been re-added.

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -540,9 +541,8 @@ void slowControl_DragDrop(object sender, DragEventArgs e)
             bool result = false;
-
-            string strSC = getSC();
-
```
Re-added this line

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -540,9 +541,8 @@ void slowControl_DragDrop(object sender, DragEventArgs e)
-
-            string strSC = getSC();
-
+            string configNo = "";
```
Moved the declaration. The configNum cannot be a byte as it is sent via sendSC() that does not accept byte as input (the configuration number is appended to strSC which forms the input for sendSC()).

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -89,8 +89,9 @@ public partial class Citiroc : Form
         int sc_triggerPolarity;
         int sc_enNor32TOc;
         int sc_enTriggersOutput;
+        int sc_sendToNVM;
```
Removed this and used the member instead.

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -1398,25 +1462,75 @@ private void button_saveSC_Click(object sender, EventArgs e)
 
         private void button_loadSC_Click(object sender, EventArgs e)
```
Re-added select NVM button and removed all select functionality from load SC

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +568,50 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                                {
+                                    configNo = Convert.ToString(Convert.ToInt32(configNo, 10), 2);
+                                    configNo = configNo.PadLeft(8, '0');
+                                    strSC = getSC();
```
removed strSC = getSC();

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +568,50 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                                }
+                                else
+                                {
+                                    configNo = Convert.ToString(Convert.ToInt32(configNo, 10), 2);
```
Form cannot return a byte because sendSC will not work properly and the implementation would have to change considerably to accomodate.

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -786,6 +836,20 @@ private void checkBox_bypassPd_CheckedChanged(object sender, EventArgs e)
             sc_bypassPd = SC.Switch((CheckBox)sender, button_sendSC);
         }
 
+        private void checkBox_sendToNVM_CheckedChanged(object sender, EventArgs e)
```
Removed event.

Joarfo commented on CitirocUI/SlowControl.cs:
```diff
@@ -89,8 +89,9 @@ public partial class Citiroc : Form
+        int sc_sendToNVM;
 
-        private static string strDefSC = "111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011111111111111111111111111111111111111111111101111110011111110111100111101110000000110000000110000000 ...
+        private static string strDefSC = "111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011101110111011111111111111111111111111111111111111111111101111110011111110111100111101110000000110000000110000000 ...
```
Reverted string to previous declaration.

tstana commented on CitirocUI/Citiroc.cs:
```diff
@@ -1266,7 +1264,10 @@ private void tabPage_calibDac_MouseEnter(object sender, EventArgs e)
                 " so every SiPM has the same gain. Input DAC can be disabled per individual channel thanks to the checkboxes.";
         }
-
+        private void button_selectSC_MouseEnter(object sender, EventArgs e)
+        {
+            label_help.Text = "Select which loaded configuration on the NVM to select.";
+        }
```
Please move to `Firmware.cs` (the button should be part of the Firmware tab, as indicated previously).

Joarfo replied:
I don't think moving the code to Firmware.cs makes much sense since no other UI element functions in the firmware tab are written here. I kept it in SlowContol.cs for now, but it could be moved to Citiroc.cs since the other firmware UI elements are there, such as button_ClearArduinoSD_Click().

tstana replied:
Right, good point! I now remembered looking at the code of `Firmware.cs` that (misleading as it may be) this file contains another class which actually handles communication to the FPGA firmware running on the Citiroc Eval Kit...

tstana commented on CitirocUI/Connect.cs:
```diff
@@ -584,6 +587,8 @@ private void comboBox_SelectConnection_SelectedIndexChanged(object sender, Event
                 // Adjust other controls
                 label_DataFile.Visible = true;
                 button_SelectNumBinsCubes.Visible = true;
+                button_selectSC.Visible = true;
```
This is unnecessary; once you move the **Select NVM** button to the _Firmware_ pane, the code handles the inactivation of all controls under the _Proto-CUBES_ tab under the _Firmware_ pane.

tstana commented on CitirocUI/Connect.cs:
```diff
@@ -548,6 +548,9 @@ private void comboBox_SelectConnection_SelectedIndexChanged(object sender, Event
                 // Adjust other controls
                 label_DataFile.Visible = false;
                 button_SelectNumBinsCubes.Visible = false;
+                button_selectSC.Visible = false; 
```
This is unnecessary; once you move the **Select NVM** button to the _Firmware_ pane, the code handles the inactivation of all controls under the _Proto-CUBES_ tab under the _Firmware_ pane.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +567,55 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                    {
+                        try
+                        {
+                            if (Convert.ToInt32(configNum, 10) > 255)
+                            {
+                                MessageBox.Show("Config number too large! Please choose a number below 255"
+                                + Environment.NewLine,
+                                "Error",
+                                MessageBoxButtons.OK,
+                                MessageBoxIcon.Error);
+                            }
```
In my view, the solution to making the check in the form would have been as easy as moving this code over to `InputForm.cs`. But I can understand if you're swamped with other activities and can make this change myself.

Joarfo replied:
Indeed that was a solution. Now the same functionality is implented in the InputForm(). I was thinking that ideally, the form should remain open while input is invalid, and was unable to get that to work.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -589,6 +637,56 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                    return;
+            }
+
+            if (InputForm.InputBox("Select Config Number on NVM", "Config Number:", ref configNum) == DialogResult.OK)
+            {
+                try
+                {
+                    if (Convert.ToInt32(configNum, 10) > 255)
+                    {
+                        MessageBox.Show("Config number too large! Please choose a number below 255"
+                        + Environment.NewLine,
+                        "Error",
+                        MessageBoxButtons.OK,
+                        MessageBoxIcon.Error);
+                    }
+                    else
+                    {
+                        byte[] citiConf = new byte[1];
+                        configNum = Convert.ToString(Convert.ToInt32(configNum, 10), 2);
+                        configNum = configNum.PadLeft(8, '0');
+                        uint intCmdTmp = Convert.ToUInt32(configNum, 2);
+                        citiConf[0] = Convert.ToByte(intCmdTmp);
+                        protoCubes.SendCommand(ProtoCubesSerial.Command.SelectNVMCitirocConf, citiConf);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    MessageBox.Show("Input of incorrect type!"
+                    + Environment.NewLine
+                    + Environment.NewLine
+                    + "Error message:"
+                    + Environment.NewLine
+                    + ex.Message,
+                    "Error",
+                    MessageBoxButtons.OK,
+                    MessageBoxIcon.Error);
+                }
+            }
+        }
+
```
This code should be moved over to `Firmware.cs`.

Joarfo replied:
I think keeping it in SlowControl.cs or moving it to Citiroc.cs makes more sense.

tstana replied:
You're right. Let us move it to `Citiroc.cs`, right after the code of `button_ClearArduinoSD_Click()`.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -589,6 +637,56 @@ private void button_sendSC_Click(object sender, EventArgs e)
+
+            if (connectStatus != 1)
+            {
+                    button_sendSC.BackColor = Color.LightCoral;
```
`button_selectSC` (otherwise the **Send SC** button gets colored when in fact the **Select NVM** button should be).

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -589,6 +637,56 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                        uint intCmdTmp = Convert.ToUInt32(configNum, 2);
+                        citiConf[0] = Convert.ToByte(intCmdTmp);
+                        protoCubes.SendCommand(ProtoCubesSerial.Command.SelectNVMCitirocConf, citiConf);
+                    }
```
Add coloring of the button green in case of success.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -1397,7 +1501,7 @@ private void button_saveSC_Click(object sender, EventArgs e)
 
         private void button_loadSC_Click(object sender, EventArgs e)
-        {
+        {            
```
Unnecessary spaces at the end of this function.

tstana reviewed (changes requested):
@Joarfo almost there. I've added some more comments. Mainly, the **Select NVM** button should be part of the _Firmware_ pane, as I mentioned previously.

If you do not have time to make these changes, I can do it for you. I anyway would need to test before merging to `master` (I'm guessing you would also not have time for that).

Joarfo commented:
I addressed all comments left here. I am wondering about where to put button_selectSC_click(), since it feels out of place in Firmware.cs since no other UI elements are written there. SlowControl.cs or Citiroc.cs would be more fitting I think. As for testing, I don't think I will have time to test the latest commit, but I did test right after I re-added the checkbox, and it worked then.

tstana commented on CitirocUI/InputForm.cs:
```diff
@@ -0,0 +1,88 @@
+                    MessageBoxIcon.Error);
+                    dialogResult = DialogResult.Cancel;
+                }
+                else
```
... remove this `else` here...

tstana commented on CitirocUI/InputForm.cs:
```diff
@@ -0,0 +1,88 @@
+
+            DialogResult dialogResult = form.ShowDialog();
+
+            try
```
This `try` statement can be put in the context of a `while` loop, e.g.:

`while (excep)`

tstana commented on CitirocUI/InputForm.cs:
```diff
@@ -0,0 +1,88 @@
+            {
+                if (Convert.ToInt32(textBox.Text, 10) > 255)
+                {
+                    MessageBox.Show("Config number too large! Please choose a number below 255"
+                    + Environment.NewLine,
+                    "Error",
+                    MessageBoxButtons.OK,
+                    MessageBoxIcon.Error);
+                    dialogResult = DialogResult.Cancel;
```
here we could do:

```C#
throw new SomeTypeOfException("Error message...");
```

**TODO:** Find the right exception to throw from [this list](https://docs.microsoft.com/en-us/dotnet/api/system.exception?view=net-6.0#choosing-standard-exceptions). For example, `ArgumentException` might be a good one to pick.

tstana commented on CitirocUI/InputForm.cs:
```diff
@@ -0,0 +1,88 @@
+                        "Error",
+                        MessageBoxButtons.OK,
+                        MessageBoxIcon.Error);
+                dialogResult = DialogResult.Cancel;
```
... and remove this guy here.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +567,29 @@ private void button_sendSC_Click(object sender, EventArgs e)
+                        strSC += strRev(configNum);
+                     }
+                }
+                else
+                {
+                    label_help.Text = "Please configure your connection.";
+                }
+                
```
This part of the text should be removed, as it is informing the user that they should configure their connection, when they are simply attempting to send a config to the ASIC.

tstana reviewed (changes requested):
@Joarfo thank you for your patience to add yet more changes!

You can find some suggestions in-line in the input form on how one could keep the user in the form until they type in a correct value.

One problematic item when testing is that the config cannot be sent to the NVM, as there is a check inside `sendSC()` for the length of the config, which fails when we add our own config. The culprit lines within `sendSC()` are:

```c#
            // Get standard length of slow control bitstream
            int scLength = strDefSC.Length;
            // Get length of current slow control bitstream
            int intLenStrSC = strSC.Length;
            byte[] bytSC = new byte[scLength / 8];
            // If the length of the current bitstream is not OK, return false
            // else store the slow control in byte array
            if (intLenStrSC == scLength)
            {
```

This check can safely be removed, as far as I can see. At the same time, the following should be removed:

- [ ] the definition of the variable `scLength` (use `strSC.Length` in its stead);
- [ ] the definition of the variable `intLenStrSC`.

@Joarfo, will you be able to make these changes and come to the office to try them out on the hardware?

tstana commented on CitirocUI/InputForm.cs:
```diff
@@ -0,0 +1,88 @@
+
+            try
+            {
+                if (Convert.ToInt32(textBox.Text, 10) > 255)
```
Use `Convert.ToUInt32` here, as this allows for the possibility of the user typing in a negative number.

Joarfo commented:
I redid the entire ConfigIdInputForm using Visual Studios tools. Now it checks for correct data type and correct size without removing the user from the form. I also removed the unnecessary check from the sendSC() function. I could come in and test this new command sometime this week probably. I need to return the key anyway.

tstana commented on CitirocUI/CitirocUI.csproj:
```diff
@@ -99,6 +99,15 @@
     <Compile Include="Citiroc.Designer.cs">
       <DependentUpon>Citiroc.cs</DependentUpon>
     </Compile>
+    <Compile Include="ConfigInputForm.cs">
+      <SubType>Form</SubType>
+    </Compile>
```
The `ConfigInputForm.cs` file needs to be removed from the project, otherwise a build error occurs (on my side at least).

tstana commented on CitirocUI/ConfigIdInputForm.cs:
```diff
@@ -0,0 +1,53 @@
+            {
+                if (Convert.ToUInt32(textBox.Text, 10) > 255)
+                {
+                    label.Text = "Input to large! Must be below 255";
```
Input _too_ large (one 'o' missing).

tstana replied:
Solved in a newer version of `ConfigIdInputForm`.

tstana commented on CitirocUI/SlowControl.cs:
```diff
@@ -568,6 +567,26 @@ private void button_sendSC_Click(object sender, EventArgs e)
             }
             else if (comboBox_SelectConnection.SelectedIndex == 1)
             {
+                if ((checkBox_sendToNVM.Checked == true) & (connectStatus != 1))
```
Logical AND needed here, i.e., `&&`.

tstana reviewed (changes requested):
@Joarfo, I think it will work now, looking forward to hear the results of the testing!

Please see the remaining comments for things still to be addressed.

tstana reviewed (approved):
@Joarfo, it looks good now! Once you confirm it's working after testing, we can merge this PR.
